From e0dd01b536919d5407968eae341b72fa10ec0b7d Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=C5=81ukasz=20Rymanowski?= <rlukasz@google.com>
Date: Thu, 26 May 2022 09:29:13 +0200
Subject: [PATCH] l2cap: Verify number of channels in the CoC request

If remote side sends request for more than 5 channels, Android shall
reject it.

Bug: 232256974
Test: atest BluetoothInstrumentationTests
Change-Id: I85b8d8358123b1713d5c6d0c4251352a7de8fa8a
(cherry picked from commit d24b9c262039347b30c1ec41d38fa5c598a9fbd7)
Merged-In: I85b8d8358123b1713d5c6d0c4251352a7de8fa8a
---
 stack/l2cap/l2c_ble.cc | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/stack/l2cap/l2c_ble.cc b/stack/l2cap/l2c_ble.cc
index d5486794d..c878ef2a5 100644
--- a/stack/l2cap/l2c_ble.cc
+++ b/stack/l2cap/l2c_ble.cc
@@ -26,6 +26,7 @@
 
 #include <base/logging.h>
 #include <base/strings/stringprintf.h>
+#include <log/log.h>
 
 #include "bt_target.h"
 #include "bta_hearing_aid_api.h"
@@ -515,6 +516,15 @@ void l2cble_process_sig_cmd(tL2C_LCB* p_lcb, uint8_t* p, uint16_t pkt_len) {
 
       /* Check how many channels remote side wants. */
       num_of_channels = (p_pkt_end - p) / sizeof(uint16_t);
+      if (num_of_channels > L2CAP_CREDIT_BASED_MAX_CIDS) {
+        android_errorWriteLog(0x534e4554, "232256974");
+        LOG_WARN("L2CAP - invalid number of channels requested: %d",
+                 num_of_channels);
+        l2cu_reject_credit_based_conn_req(p_lcb, id,
+                                          L2CAP_CREDIT_BASED_MAX_CIDS,
+                                          L2CAP_LE_RESULT_INVALID_PARAMETERS);
+        return;
+      }
 
       LOG_DEBUG(
           "Recv L2CAP_CMD_CREDIT_BASED_CONN_REQ with "
-- 
2.37.0.rc0.161.g10f37bed90-goog

